pandas/core/internals/blocks.py

```diff
@@ -882,6 +882,14 @@ def setitem(self, indexer, value):
             and arr_value.size == values.size
         ):
             values[indexer] = value
+
+            if self.is_categorical_astype(arr_value.dtype) and not is_categorical_dtype(
```

Comment:
I led you astray a bit here, we should do this as part of the elif, so put this on line 877 (e.g. before *this* elif)

Reply:
you *might* be able to do this even higher in the function (which is pretty long / complex).

Reply:
Moved it to part of elif.